pkg/entity/tournament.go

```diff
@@ -0,0 +1,769 @@
+package entity
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/rand"
+	"sort"
+	"sync"
+	// "time"
+	"gorm.io/datatypes"
+
+	realtime "github.com/domino14/liwords/rpc/api/proto/realtime"
+)
+
+type Result int
+
+const (
+	None Result = iota
+	Win
+	Loss
+	Draw
+	Bye
+	ForfeitLoss
+	ForfeitWin
+	Eliminated
+)
+
+type TournamentGame struct {
+	Scores        []int
+	Results       []Result
+	GameEndReason realtime.GameEndReason
+}
+
+const Unpaired = -1
+
+type Pairing struct {
+	Players  []string
+	Games    []*TournamentGame
+	Outcomes []Result
+}
+
+type PlayerRoundInfo struct {
+	Pairing *Pairing `json:"p"`
+	Record  []int    `json:"r"`
+	Spread  int      `json:"s"`
+}
+
+type Standing struct {
+	Player string
+	Wins   int
+	Losses int
+	Draws  int
+	Spread int
+}
+
+type Tournament interface {
+	GetPlayerRoundInfo(string, int) (*PlayerRoundInfo, error)
+	SubmitResult(int, string, string, int, int, Result, Result, realtime.GameEndReason, bool, int) error
+	GetStandings(int) ([]*Standing, error)
+	SetPairing(string, string, int) error
+	IsRoundComplete(int) (bool, error)
+	IsFinished() (bool, error)
+	Serialize() (datatypes.JSON, error)
+}
+
+type PairingMethod int
+
+const (
+	Random PairingMethod = iota
+	RoundRobin
+	KingOfTheHill
+	Elimination
+	// Need to implement eventually
+	// Swiss
+	// Performance
+
+	// Manual simply does not make any
+	// pairings at all. The director
+	// has to make all the pairings themselves.
+	Manual
+)
+
+type TournamentClassic struct {
+	Matrix         [][]*PlayerRoundInfo `json:"t"`
+	Players        []string             `json:"p"`
+	PlayerIndexMap map[string]int       `json:"m"`
+	PairingMethods []PairingMethod      `json:"m"`
+	GamesPerRound  int                  `json:"g"`
+}
+
+type TournamentType int
+
+const (
+	ClassicTournamentType TournamentType = iota
+	// It's gonna be lit:
+	ArenaTournamentType
+)
+
+type TournamentManager struct {
+	sync.RWMutex
+	UUID string `json:"u"`
+
+	Name       string         `json:"n"`
+	Directors  []string       `json:"d"`
+	Type       TournamentType `json:"t"`
+	Tournament Tournament     `json:"m"`
+}
```

Comment:
we can also add a Description text field so that this can show when you click on the tournament. Kinda like a flyer with the rules and stuff.

Reply:
And a date for the tournament

Reply:
Done